torch/csrc/api/include/torch/data/ctf/ctf_parser.h

```diff
@@ -0,0 +1,931 @@
+    char c = reader_->peek_char();
+    if (!is_digit(c)) {
+#ifdef CTF_DEBUG
+      std::cout << "Not a Sequence ID at position " << initial_pos << std::endl;
+#endif
+      if (has_initial_sequence_id_) {
+        sequence_id = previous_sequence_id_;
+#ifdef CTF_DEBUG
+        std::cout << "Using previous Sequence ID (" << previous_sequence_id_
+                  << ")" << std::endl;
+#endif
+      } else {
+        is_new = true;
+        sequence_id = previous_sequence_id_ + 1;
+
+#ifdef CTF_DEBUG
+        std::cout << "Incremented previous Sequence ID (" << sequence_id << ")"
+                  << std::endl;
+#endif
+      }
+      previous_sequence_id_ = sequence_id;
+      return is_new;
+    }
+
+    // Get all consecutive digits
+    while (is_digit(reader_->peek_char())) {
+      c = reader_->get_char();
+      scratch_[idx++] = c;
+    }
+    scratch_[idx] = '\0';
+
+    // Discard delimiters after the ID
+    while (is_value_delimiter(reader_->peek_char())) {
+      reader_->get_char();
+    }
+
+    // After Sequence ID, there must be a '|'
+    if (!is_name_prefix(reader_->peek_char())) {
+      std::string error_msg(
+          "Missing name delimiter for one of the sequences at position " +
+          std::to_string(reader_->get_position()));
+#ifdef CTF_DEBUG
+      std::cerr << error_msg << std::endl;
+#endif
+      throw std::runtime_error(error_msg);
+    }
+
+    // Convert string and return integral value
+    sequence_id = static_cast<CTFSequenceID>(std::stoull(scratch_.data()));
+#ifdef CTF_DEBUG
+    std::cout << "Found Sequence ID '" << std::to_string(sequence_id)
+              << "' at position " << std::to_string(initial_pos) << std::endl;
+#endif
+
+    // Decides whether this is a new example or an existing one
+    if (previous_sequence_id_ != sequence_id && sequence_id != LONG_MAX) {
+      is_new = true;
+    }
+
+    previous_sequence_id_ = sequence_id;
+    has_initial_sequence_id_ = true;
+    return is_new;
+  }
+
+  bool get_input_stream(
+      const CTFSequenceID& sequence_id,
+      bool& is_new_sequence) {
+    // Create a new sequence with input_streams_info pre-allocated
+    if (is_new_sequence) {
+      is_new_sequence = false;
+
+#ifdef CTF_DEBUG
+      dataset_->sequences_id.emplace_back(sequence_id);
+#endif
+
+      // New sequence to be appended to dataset_->sequences
+      CTFSequenceData sequence;
+
+      for (auto const& stream : dataset_->input_streams_info) {
+        CTFInputStreamID input_stream_id =
+            dataset_->input_streams_map[stream.name];
+        if (stream.storage == CTFDataStorage::Dense) {
+          // TODO: Performance consideration: CNTK knows the number of samples
+          // in the sequence, allowing accurate memory reservation (index built
+          // during init)
+          sequence.emplace_back(
+              std::make_shared<CTFDenseInputStreamData<DataType>>(
+                  input_stream_id, stream.dimension));
```

Comment:
Yup, the reason for that is that an input stream is a multiple of stream.dimension

0 |a 1 2 3 |b 4
0 |a 4 5 6 |b 5

a has dimension 3, but the input stream should be 6, because it is comprised of two samples, 3 values long each

b has dimension 1, but the input stream should be 2 for the same reason

Reply:
That is why it is difficult to reserve memory, even for dense input stream. we dont know how many examples there are in a sequence for each input stream